### ISSUE TITLE
Downsampling / Deduplication per label / tenant

### ISSUE THREAD

**Issue by tenmozes** [VictoriaMetrics/VictoriaMetrics#4960]
### Is your feature request related to a problem? Please describe

I'd like to have different downsampling config per label/tenant.

### Describe the solution you'd like

Introduce an option to define combine time series selector and downsampling period in one of the following ways:
1. Update the existing `-downsampling.period=0s:D` flag and introduce the optional filter. Example:  `-downsampling.period={instance!="bar"}:5s:10d`. The notation without filter is equivalent to `-downsampling.period={__name!=""}:5s:10d`
2. Add `-downsampling.filter` option to specify the filter with the downsampling period. So global settings are in the `-downsampling.period` flag, by label/filter in `-downsapmling.filter`
3. Add `-downsampling.filter` where you can specify filter only, the period still be in `-downsampling.period`, the array position must be taken into account (like with other flags that support array as a parameter and depend on other flags)

Possible scenarios that may be covered 
- [ ] Global downsampling and downsampling per tenant set. For the same periods/intervals, the expected behavior filter must be applied. Example `downsampling.period=5s:10d` and `-downsampling.period={instance!="bar"}:10s:10d`
- [ ] Global downsampling and downsampling per tenant set. For the different periods/intervals, the expected behavior -  downsampling of all or a subset of the metrics (it depends which global/filter position) is applied. Example
`downsampling.period=5s:10d` and `-downsampling.period={instance!="bar"}:10s:20d`

The tenant can be used as the label (reuse the current approach from multitenancy by label feature). Example `-downsampling.period={vm_account_id="42"}:10s:20d`

Update1: - There are a new use cases
- exclude some time series from downsampling 
- use downsampling for deduplication per tenant for all metrics

**Comment by PatMis16** [comment]
Can I vote for this?

**Comment by hagen1778** [comment]
Hello @tenmozes @valyala!
I like the p1 from @tenmozes where he suggests extending the existing downsampling cmd-line flag. When reading this I tried to use [retention filters](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#retention-filters) as a reference. But the difference between downsampling and retentionPeriod is that the latter is a required param. We always should have `retentionPeriod` set as default, as a fallback if none of `retentionFilter` selectors matched. 

But downsampling is optional, so maybe we can define the following rule for downsampling config:
1. It is allowed to pass multiple `downsampling.period` flags
2. Each `downsampling.period` should contain time series selector.
3. If time series selector is omitted, it is automatically filled with `{__name__!=""}` selector. In other words, it automatically matches everything.
4. The order of passed `downsampling.period` matters, since only first matched selector will be applied.

See some examples below:

| Scenario                                                                                          | Config example                                                                                                         |
|---------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------|
| Downsample all data                                                                               | `-downsampling.period=1m:10d` or `-downsampling.period={__name__!=""}:1m:10d`                                          |
| Downsample only a specific selector, <br/>the rest of data remains untouched                      | `-downsampling.period={team="dev"}:1m:10d`                                                                             |
| Downsample multiple selectors,<br/>the rest of data remains untouched                             | `-downsampling.period={team="dev"}:1m:10d`, `-downsampling.period={team="prod"}:1m:90d`                                |
| Downsample specific tenants,<br/>the rest of tenants remains untouched                            | `-downsampling.period={vm_account_id="0"}:1m:10d`,  `-downsampling.period={vm_account_id="2"}:1m:90d`                  |
| Downsample all but a specific selector                                                            | `-downsampling.period={team!="prod"}:1m:10d`                                                                           |
| Downsample all but multiple specific selectors                                                    | `-downsampling.period={team!~"(prod&#124;staging)"}:1m:10d`                                                            |
| Downsample `team=dev` after 30d, `team=prod` after 90d, all the rest after 10d                    | `-downsampling.period={team="dev"}:1m:30d`, `-downsampling.period={team="prod"}:1m:90d`, `-downsampling.period=1m:10d` |
| Downsample all after 10d. <br/>Please note, the order is important: only first matcher is applied | `-downsampling.period=1m:10d`, `-downsampling.period={team="dev"}:1m:30d`, `-downsampling.period={team="prod"}:1m:90d` |

Please, let me know what you think.

**Comment by tenmozes** [comment]
@hagen1778, how to downsample all data expect specific selectors? To be more specific how to exclude data from downsampling

**Comment by tenmozes** [comment]
is it possible to setup downsampling after 0s (deduplication per tenant)?

**Comment by valyala** [comment]
@hagen1778 , I like the proposed configuration for `-downsampling.period`! Let's implement it!

> is it possible to setup downsampling after 0s (deduplication per tenant)?

Yes, just specify `-downsampling.period={vm_account_id=~"42|99|123"}:1m:0s` for applying the 1 minute deduplication only to the specified tenants - `42`, `99` and `123`.

**Comment by hagen1778** [comment]
One of the questions we need to either discuss or accept is the aggregation of time series with different downsampling period. For example, `-downsampling.period={team="dev"}:5m:10d` `-downsampling.period={team="prod"}:1m:10d` - this instructs VM to apply downsampling to all time series with `team="dev"` to 5m and for time series with `team="prod"` labels to 1m after 10d. If user will try to query something like `sum(rate(my_counter[5m]))` without specifying filter by `team` then VM will have to aggregate time series with different resolution, 1m and 5m. This may produce unexpected results as time series aren't aligned.
What can be done:
1. Explain the problem to the user via docs, so user will reduce the probability of such intersections between series with different downsampling period
2. Try to figure on query time the biggest downsampling period and apply it to all time series in request. This may require additional computation resoruces and will increase query latency

I think p1 is preferable since neither Prometheus nor VictoriaMetrics restricts having scrape targets with different scrape_intervals. I don't remember many complains regarding aggregation of non-aligned time series. It could be because:
a. People mostly do not care or do not notice such discrepancies
b. People are very serious about setting consistent scrape_interval.

**Comment by valyala** [comment]
There is an implementation obstacle, which needs to be resolved before this feature will be implemented - VictoriaMetrics applies the downsampling interval globally per each `part` stored on disk (see [these docs](https://docs.victoriametrics.com/#storage) for details on what is a `part`), and stores the applied downsampling interval at `metadata.json` files inside the part directory. This allows skipping applying the same or smaller downsampling interval to the same part multiple times. It is unclear yet how to deal with multiple downsampling intervals, which can be applied to different time series stored in the same `part`. @f41gh7 , probably you have ideas on how to resolve this question?

**Comment by hagen1778** [comment]
> It is unclear yet how to deal with multiple downsampling intervals, which can be applied to different time series stored in the same part

Hey @valyala! WDYT about adding a new field to metadata.json to contain unique applied downsampling filters to this specific part? See the structure below:
```
{
  "RowsCount": 444309,
  "BlocksCount": 2918,
  "MinTimestamp": 1700045821173,
  "MaxTimestamp": 1700046261173,
  "MinDedupInterval": 0, // Global dedup setting
  "MinDedupIntervalFilters": [ // New field with dedup filters
    '{team="dev"}:1m:30d',
    '{team="prod"}:1m:90d'
  ]
}
```
`MinDedupInterval` could be deprecated to become `"MinDedupIntervalFilters": ['{__name__!=""}']`, but it could be better to save it as is for backward-compatibility purposes.

The new `MinDedupIntervalFilters` field, similar to `MinDedupInterval`, works as a "mark" on data part to reflect whether specific downsampling rule was already applied. The rationale is the following:
1. Data part contains a set of all time series ingested on MinTimestamp-MaxTimestamp period, so it is safe to assume that MinDedupIntervalFilters will match at least something in the part. At least, it looks more performance-wise to assume so instead of actually locating if a matched time series belongs to this part. But I could be wrong about this.
2. Once part was deduplicated, it gets a MinDedupIntervalFilters mark and can be skipped on next iteration.
3. Unless, `{team="dev"}:1m:30d` has been changed to bigger interval {team="dev"}:**5m**:30d" or period "{team="dev"}:1m:**90d**". If interval or period were changed to smaller values - we skip such part during downsampling as there is nothing we can do about it.
4. Adding a new `--downsampling.period={filter}:interval:period` will schedule all parts to be downsampled/deduplicated again, as they might contain matched series.

**Comment by valyala** [comment]
Filter-based downsampling is available in VictoriaMetrics starting from [v1.100.0](https://docs.victoriametrics.com/changelog/). See [these docs](https://docs.victoriametrics.com/#downsampling) for details. Per-tenant downsampling in VictoriaMetrics cluster is also available starting from `v1.100.0` according to [these docs](https://docs.victoriametrics.com/cluster-victoriametrics/#downsampling).

Closing this feature request as done.

**Comment by akenO8** [comment]
> Filter-based downsampling is available in VictoriaMetrics starting from [v1.100.0](https://docs.victoriametrics.com/changelog/). See [these docs](https://docs.victoriametrics.com/#downsampling) for details. Per-tenant downsampling in VictoriaMetrics cluster is also available starting from `v1.100.0` according to [these docs](https://docs.victoriametrics.com/cluster-victoriametrics/#downsampling).
> 
> Closing this feature request as done.

Is that available in opensouce version?

**Comment by hagen1778** [comment]
@akenO8, [downsampling](https://docs.victoriametrics.com/#downsampling) is an Enterprise feature.
